src/hooks/queries/useToggleCartMutation.ts

```diff
@@ -0,0 +1,36 @@
+import addProductToCart from '@/apis/addProductToCart';
+import removeProductFromCart from '@/apis/removeProductFromCart';
+import { Product } from '@/models/product.model';
+import { useMutation, useQueryClient } from '@tanstack/react-query';
+import { PRODUCTS_QUERY_KEY } from './useProductsQuery';
+
+interface UseToggleCartMutationRequest {
+  productId: number;
+  isInCart: boolean;
+}
+
+export default function useToggleCartMutation({ productId, isInCart }: UseToggleCartMutationRequest) {
+  const queryClient = useQueryClient();
+
+  return useMutation({
+    mutationFn: isInCart ? () => removeProductFromCart(productId) : () => addProductToCart(productId),
+    onMutate: async () => {
+      await queryClient.cancelQueries({ queryKey: [PRODUCTS_QUERY_KEY] });
+      const previousProducts = queryClient.getQueryData([PRODUCTS_QUERY_KEY]);
+
+      queryClient.setQueryData([PRODUCTS_QUERY_KEY], (old: Product[]) =>
+        old.map((item) => {
+          if (item.id === productId) {
+            return { ...item, isInCart: !item.isInCart };
+          }
+          return item;
+        })
+      );
+
+      return { previousProducts };
+    },
+    onError: (_, __, context) => {
+      queryClient.setQueryData([PRODUCTS_QUERY_KEY], context?.previousProducts);
```

Comment:
```suggestion
if(context) {
    queryClient.setQueryData([PRODUCTS_QUERY_KEY], context.previousProducts);     
}
```
onMutate의 반환값이 있어서 플로우상 우려가 되진 않지만 옵셔널 체이닝 사용시 `해당 부분이 undefined 값으로 사용되어도 괜찮은가?`는 항상 고민해보는게 좋을 것 같아요.

Reply:
오 좋은 포인트네요. undefined값이 사용되면 안되고 메서드의 타입을 맞춰주고 있을 뿐이니 옵셔널 체이닝을 붙일 이유가 없는 것 같아요

Reply:
https://github.com/next-step/react-shopping-cart/pull/38/commits/3a312f9fb7cd47895a1f81c693e6ef9e8431e8a6